### ISSUE TITLE
Unimplemented HLE wrt Networking

### ISSUE THREAD

**Issue by waveywaves** [hrydgard/ppsspp#12542]
Considering all the unimplemented functions in HLE
https://github.com/hrydgard/ppsspp/blob/aa927e0681ed76fec3b540218981a2c407a78e47/Core/HLE/sceNet.cpp#L230-L259
Is there a timeline regarding when these would be actively supported ? 
### What happens?
```
06:17:015 user_main    E[SCENET]: HLE/sceNet.cpp:397 UNIMPL sceNetInetSocket(2, 1, 0)
```

### What *should* happen?
It should work

### What hardware, operating system, and PPSSPP version?  On desktop, GPU matters for graphical issues.

```
OS : Fedora
PPSSPP Version : latest
```

**Comment by unknownbrackets** [comment]
It's waiting for you or someone to come along and start implementing them properly.

IMHO, that means creating tests with [pspautotests](https://github.com/hrydgard/pspautotests), and probably in some cases requires running the tests on two PSPs concurrently (and being able to execute the tests in PPSSPP the same way.)

-[Unknown]

**Comment by waveywaves** [comment]
@unknownbrackets it would be nice to contribute on this. Currently I don't have a PSP, but can I start without having to buy one. I went through https://www.ppsspp.org/development.html Didn't get a clear idea of what is necessary in terms of contribution to HLE.

**Comment by unknownbrackets** [comment]
To get started at first, you need a way to compile PPSSPP.  This is pretty easy - on Windows, you'll need git, Visual Studio, and python 3.  On Linux, you'll need git, cmake, python 3, and clang.  On macOS, you'll mostly just need Xcode and its command line tools.

When PSP games call HLE functions, they set registers, execute a `syscall` instruction, and expect the result in a register.  But from your perspective, you can mostly ignore this.  When you see:

```c++
int sceSomeFunc(int a, int b) {
    return 4;
}
```

That's already handling the registers for you - there's actually a wrapper that translates the return `int` and the `a` and `b` parameters to/from registers magically for you.

The next thing to understand is how execution works.  Basically, PSP games can start multiple threads - but they cannot execute concurrently (just as on a real PSP.)  So while inside your `sceFooBar` function, nothing else is happening.  If you `usleep()` inside the function, it just means emulation is paused for that time.  So you basically can't do that.

After that, I'd suggest peeking at some `DoState` functions.  Basically, this is how save states work: there's a single function that is run which either writes (serializes) a save state, or reads it, but the code is the same.  In some cases the `mode` is checked.  It's important that any state which gets introduced be included in the save state.  We also have versioning so that backwards compatibility of save states can be managed.

That's enough of a primer to make changes to HLE functions.  You can try guess work - change the function and see if doing that helps things.

But the way I have done it for other HLE functions more closely resembles TDD, and unfortunately largely requires a PSP.  Basically, I write a small program for the PSP, and it tries various different parameters for a function - for example, maybe calling an `sceFooInit` function twice, and checking if it should return an error code the second time.  Then, I run this same program under PPSSPP, and modify PPSSPP until it gives the same output as a real PSP did.

Sometimes the most time in this process is spent writing the test program, to figure out even what questions should be asked about how the function works.  Once I have a picture of how it works from the outputs it gives, it's usually not too complex to model it.

I've had limited time of late, but if you want to try writing a test program, I can run it for you.  I've also toyed with the idea of trying to setup a simple network tool so that users could connect their old PSPs to form a sort of "cloud" to execute these tests, to make it easier for people who want to contribute to PPSSPP without their own actual PSP.

-[Unknown]

**Comment by adenovan** [comment]
"PSPs to form a sort of "cloud" to execute these tests" this is the interesting part from this issues.

if there is a way to execute the test through cloud. it is doable to execute the game through network with same controller input? i know you have implement something interesting like remote iso if PSP can do that also only the remote iso things and we can inspect the actual function call returned from network it will be much easier to see the function model and implement it on HLE. 

it will be a better test suite cause we inspect two different implementation at same time the real PSP implementation and emulator HLE result in real time. 

lately i have been encounter some games strange behavior looks like the game call HLE function like this 

`
  while(networkInited && peerInList < 4 ){
    rc= SceNetAdhocPtpAccept(sock,peerMac,peerPort,timeout,flag);
    if(rc == error){ 
      waitForSomethingThatNeverThere();
    }
  }
`
^ this game repeatedly spam the hle function if peer less than the condition, we cannot break the loop if we return wrong error or even a success call. the emulation is still running but waiting for something that never there. guessing not much help there and write a test program to do that behavior is really complex because there is so many different params that game send. returning different error make the the HLE call making another HLE call agains.

so the best way to emulate this is inspect the game on actual console itself with many test condition or another psp hardware to test. if we can MITM the game when its connected to the network and peer condition is meet with another psp hardware the implementation will be close.

**Comment by unknownbrackets** [comment]
Take a look at [pspautotests](https://github.com/hrydgard/pspautotests).  We simply use usbhostfs_pc and psplink to run simple compiled prxs on the PSP.

Executing commercial games on a network of user's PSPs to test things also has a lot of problems I'm not at all prepared to worry about.  I don't think the laws in my country would allow me to contribute to such an effort, or possibly even provide advice about it.

I still think it's better to boil down usage.  An overall approach that I think works well is:

1. Start with an exhaustive test of accepted parameters.
   * What does the HLE call do with a bad sock?
   * Can a MAC be passed that is "wrong"?
   * What's the valid range of port?
   * Is timeout optional?
   * What flags are supported, and what do they do?

2. Next validate combinations of parameters that might confclit.
   * If there's a mac that's a wildcard, can it be used with a port that is a wildcard?
   * Are there flags that make certain parameters ignored, even for errors?

3. Validate stateful behavior within the function itself.
   * What if I call it twice (in this case, accept probably just gets a new sock that has connected, so less relevant here)?
   * If the sock is closed, is that different from an invalid sock?
   * Is it worth testing what happens if two threads try to accept on a single sock simultaneously (probably only if games seem to do this)?

4. Validate stateful behavior within the function family.
   * Now we finally set up a test scenario: thread A binds the socket and calls accept, thread B tries to connect to it.
   * Can we connect and accept non-blocking?  Does it queue?
   * How much of a backlog does it use on accept?
   * What happens if the network switch is disabled?
   * What if network is not inited first or is shut down before calling?

It's absolutely true that this doesn't exactly cover the problem set.  But I promise, if every single function had this treatment (like the sceKernel funcs, sceMp3, sceCcc, sceRtc, etc.) it'd be a ton easier to investigate the complex scenarios.  You'd have most of the error codes, and no what they mean.  Comparing with a JpcspTrace or similar would give you very direct ideas on what might be happening.  And there'd be way less concern that another function is behaving wrong, so this one is being called in a way that just couldn't ever work.

Staring at a certain usage of a func and trying to figure it out when every other func is basically 50% guesswork and 50% bugs won't take less time than the above.  But it is time consuming to write all the tests.

The other benefit of these tests is they can keep you from breaking things later: they're automated, like unit tests.  You just run them to check if you've broken something that used to work.  It's a lot easier than replaying a game, even if it doesn't fully replicate the entire game (or replace testing games that like to break.)

JpcspTrace is a good example of doing some MITM testing.  It just logs all calls to specified funcs.

In your specific example, I could imagine it being something like this:

```c++
registerNetworkCallback(&networkCallback);

while (networkInited && peerInList < 4) {
    rc = SceNetAdhocPtpAccept(sock, peerMac, peerPort, timeout, flag);
    if (rc == ERROR_NETWORK_UNAVAILABLE) {
      waitForEventFlag();
   }
}

void networkCallback(int ev) {
    if (ev == NETWORK_STATE_RESTORED) {
        setEventFlag();
    }
}
```

Maybe it has nothing to do with the accept func, even.

-[Unknown]

**Comment by adenovan** [comment]
yep looking at your code example it all make sense there is nothing to do with the accept function itself cause all the game that did not use AdhocMatching work so well in the current adhoc module.

More likely the problem is the game waiting for flag that never got there cause there is something broken in the matching function or the matching flag is not get there due to network failure, its happens on many games like DISSIDIA 012 , Phantasy Star Portable 2 this game open 12 socket or even accept more that make my head dizzy when looking into this strange behavior on the accept that return either success or false no matter what returned from the accept function the game never go any further.

the end result is sometime the game enter the adhoc mode or sometimes its just hang waiting something that supposed to happen but there is nothing in event queue. 

already look at that pspautotest and all have the dev environment for PSP. the only problem i have is there is no another PSP to actually test the adhoc program using all those psp adhoc apis. that strange behavior itself make me things to brought 3 more psp just to simulate all of these.

im thinking of written some simple adhoc chat homebrew to inspect PDP , PTP and Matching behavior within real hardware. so it got more clear if we do something that not working in adhoc module using all psp apis there.
